### PR TITLE
Better behaviour for using autoscale in combination with log norm

### LINKED ISSUE #1544: Rescale uses incomplete range in 2D plots?
Manual bounds, note e.g., peaks in the right half:
image

After hitting rescale:
image

Not sure if this happens only in log-scale?

### PR DESCRIPTION
**Pull request by nvaytet**
The behaviour should be the following:

- if `vmin` or `vmax` have been requested by the user, these limits stay in place, even if we zoom or toggle log norm.
- once we hit rescale, the user's initial `vmin`/`vmax` are lost
- auto-rescaling for log norm is done by histogramming the data values in bins between 1.0e-30 and 1.0e30 and finds the edges of the first and last non-zero bins
- if limits are not requested by the user, we rescale automatically on zoom

Fixes #1544

### CODE REVIEW

File: python/src/scipp/plot/tools.py [hunk 1]
```diff
@@ -117,21 +117,56 @@ def vars_to_err(v):
...
-    if lims is not None:
-        return np.array([vmin, vmax])
+def find_log_limits(x):
+    """
+    To find log scale limits, we histogram the data between 1.0-30
+    and 1.0e+30 and include only bins that are non-zero.
+    """
+    volume = np.product(x.shape)
+    pixel = sc.reshape(x, dims=['pixel'], shape=(volume, ))
+    pixel.variances = None  # no variances allowed in coord for histogram
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/tools.py:127]:
```suggestion
    pixel = sc.reshape(sc.values(x), dims=['pixel'], shape=(volume, ))
```

File: python/src/scipp/plot/tools.py [hunk 2]
```diff
@@ -117,21 +117,56 @@ def vars_to_err(v):
...
+    weights = sc.Variable(dims=['pixel'], values=np.ones(volume))
+    hist = sc.histogram(sc.DataArray(data=weights, coords={'order': pixel}),
+                        bins=sc.Variable(dims=['order'],
+                                         values=np.geomspace(1e-30,
+                                                             1e30,
+                                                             num=61),
+                                         unit=x.unit))
+    # Find the first and the last non-zero bins
+    has_data = hist.data > 0.0 * sc.units.one
+    inds = np.where(has_data.values)
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/tools.py:137]: https://numpy.org/doc/stable/reference/generated/numpy.where.html says one should use `nonzero` directly?

File: python/src/scipp/plot/tools.py [hunk 3]
```diff
@@ -117,21 +117,56 @@ def vars_to_err(v):
...
+    ar = np.arange(hist.data.shape[0])[inds]
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/tools.py:138]:
```suggestion
    ar = np.arange(volume)[inds]
```

File: python/src/scipp/plot/tools.py [hunk 4]
```diff
@@ -117,21 +117,56 @@ def vars_to_err(v):
...
+    # Safety check in case there are no positive values:
+    # return arbitrary values that ensure toggling log on the plot with buttons
+    # does not fail.
+    if len(ar) == 0:
+        vmin = 0.1
+        vmax = 1.0
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/tools.py:144]: Should we fallback to `find_linear_limits` instead? Or something slow giving the right limits? Or the old mechanism?

Comment (nvaytet) [reply python/src/scipp/plot/tools.py:142]: This is the case where there are no positive values in the data, so there are no "right" limits for a log scale.
I tried setting both to `None` but this was causing errors elsewhere (they seemed to then be overwritten somewhere else with negative values).
I can try and debug further (maybe it's better to understand fully what is happening), or keep it like this for now, which is has a stable behaviour.

Comment (nvaytet) [reply python/src/scipp/plot/tools.py:142]: actually it's not being overwritten, it's a matplotlib behaviour: if you supply `vmin=None` to a `LogNorm()` for a figure where there are only negative values, it complains that `minvalue must be positive`.

Comment (SimonHeybrock) [reply python/src/scipp/plot/tools.py:142]: What about data outside the `1e+-30` limits?

Comment (nvaytet) [reply python/src/scipp/plot/tools.py:142]: should we just extend the range above then?
you mean if someone is plotting stuff which has values between 1.0e-50 and 1.0e-30?

Comment (SimonHeybrock) [reply python/src/scipp/plot/tools.py:142]: > you mean if someone is plotting stuff which has values between 1.0e-50 and 1.0e-30?

Yes.

But by how much would you extend it? I am not sure extending can be the ultimate solution?

Comment (nvaytet) [reply python/src/scipp/plot/tools.py:142]: so then find a min and a max manually, with `sc.min` and `sc.max`, then if both are negative just return maybe 0.1 and 1.0, and if max is positive, then set min to 1.0e-3 * max?

Comment (SimonHeybrock) [reply python/src/scipp/plot/tools.py:142]: Yes, something like that 👍

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**